### ISSUE TITLE
Use Copyright Information on Youtube Videos for Track, Artist, and Album Information

### ISSUE THREAD

**Issue by Wolfrum5150** [web-scrobbler/web-scrobbler#2742]
Instead of trying to parse the title of the video, which can lead to many errors such as including information about live performances or possibly other information in the title such as a release date, us the Youtube provided copyright information on videos to fill out information for a scrobble.  

**Describe the solution you'd like**
Copyright information should take priority over video title name when deciding that information to use. This cuts down on titles with extraneous information, usually concerning live performances. Alternatively, it could be an option to turn on or off for those who prefer the current way.

**Describe alternatives you've considered**
I've considered that it may be an issue of not being able to access this information without input from the user or by needing an API. In this instance, I understand that it may not be feasible to include this feature.

**Comment by alexesprit** [comment]
Could you provide an example?

**Comment by Wolfrum5150** [comment]
> 
> 
> Could you provide an example?

Sure! In [this video of RHCP](https://www.youtube.com/watch?v=Z0AXjUy1_gY), web scrobbler currently uses this information:
![image](https://user-images.githubusercontent.com/79181893/108589587-f9b3d680-732c-11eb-866f-fbc8dcbdf618.png) 

If you open up the description, however, you can usually find something like this at the end of videos that contain copywritten music:
![image](https://user-images.githubusercontent.com/79181893/108589624-2d8efc00-732d-11eb-97c1-fb749b5f622c.png)


**It's not on all live performances, but for those that have it, it's much more standardized than the video title.**


### Some instances lose the live performance element in the copyright information, such as [this recording.](https://www.youtube.com/watch?v=jZuYXEgDJFs)
![image](https://user-images.githubusercontent.com/79181893/108589698-a5f5bd00-732d-11eb-9893-455a4f13b340.png)

### But in others, it greatly cleans up the information such as [this performance.](https://www.youtube.com/watch?v=PMzgh_1j838)

**What Web-Scrobbler records:**
![image](https://user-images.githubusercontent.com/79181893/108589743-e1908700-732d-11eb-827c-fcbf13b11c9d.png)

**What Youtube has as copyright information:**
![image](https://user-images.githubusercontent.com/79181893/108589748-eead7600-732d-11eb-83c4-f466dd931bb1.png)

**Comment by alexesprit** [comment]
Some videos contain multiple entries in this section (e.g. [this one](https://www.youtube.com/watch?v=xOraQdSTWbU)). How should we handle them?

I see two options:
1. Use information from the first entry.
2. Ignore the section, and use other sources.

I  think the second option is better - it's most likely not a music video, if there are multiple songs in the video description.

**Comment by Wolfrum5150** [comment]
I definitely agree with your idea. 

If a video has multiple entries, it could be a case like you showed, in which case, the other sources should be used.
However, in the instance of a full album video [like this](https://www.youtube.com/watch?v=-iB10tgKHkw), it only includes a limited number of tracks. In that case, the video title would definitely be prefered.

Considering that web scrobbler provides the option to only include videos in the "Music" and "Entertainment" categories, I think using the entry method for single tracks and title method for multiple tracks would help with standardization.

I definitely feel that multiple entries should resign control back to how it is now.

**Comment by alexesprit** [comment]
**TL; DR**: I found two problems while implementing this feature, though I made the basic functionality.

Current progress:
- Loading data from the copyright section is complete
- Collapsing the video description breaks the data flow (the extension will take track info from video title again)
- It should not work (because I broke it :D) with the YouTube mini player

Preview builds:

- [web-scrobbler-chrome.zip](https://github.com/web-scrobbler/web-scrobbler/files/6038906/web-scrobbler-chrome.zip)
- [web-scrobbler-firefox.zip](https://github.com/web-scrobbler/web-scrobbler/files/6038907/web-scrobbler-firefox.zip)

[How to install the extension from a zip file][install-from-zip]. Note that it will be installed as a separate extension.

[install-from-zip]: https://github.com/web-scrobbler/web-scrobbler/wiki/Install-an-unpacked-extension

<details>
  <summary><b>Read the long description of these problems</b> (Click to expand)</summary>

The current architecture is made so that the connector expects the track info is always accessible (i.e. available to get with API).

So, the two problems:
1. The copyright information is not accessible via DOM API (i.e. we can't read it in the extension) until the video description is expanded. By default, it's collapsed. So, if you expand the description with the "Show more" button, the copyright information will be loaded, and the extension can get this information.
2. When a user collapses the video description, the copyright information is unloaded after a while (about 5 seconds), and it becomes inaccessible when it's unloaded. That's how YouTube works, and we can't change this behavior.

The first problem can be solved by bringing back a hack to expand the video description programmatically (we used this hack to get video categories).

The second problem is more complex. Everything is fine, unless the video category is collapsed by the user. This could be fixed by:
a) keeping the video description expanded after using the "collapse" hack I mentioned above; but it still can be collapsed by the user.
b) saving the last track information once, and don't get the copyright info every time; so we don't care if it becomes inaccessible.

I would go with b), but it requires reworking the connector, so there could be regressions.
</details>

**Comment by Wolfrum5150** [comment]
After using the preview build, I can see what you're saying by it disappearing after the user closes.

If I'm not wrong, the preview build currently uses Method A. Shortly after the user closes the description, Web Scrobbler defaults back to the behavior we currently have now.

So Method B would be storing the track information locally? I'm just slightly lost with what you're suggesting. The way I would think of it would be to save the copyright information locally until the tab is refreshed or goes to a new page. 

I do understand that going with Method B would take a considerable amount of work and could undo some work. I would offer advice if I knew more about extension development and everything involved, but it's not anything I have experience with, so I trust that you would be able to choose the best option for development.

**Comment by alexesprit** [comment]
> If I'm not wrong, the preview build currently uses Method A. Shortly after the user closes the description, Web Scrobbler defaults back to the behavior we currently have now.

Yes, I mentioned this. I understand this is not usable, though you could test the basic functionality and leave feedback. I tested the build on different videos (e.g. with multiple copyright entries, or w/o album infor, and so on), and it works as expected, but I could miss some edge cases.

> So Method B would be storing the track information locally? I'm just slightly lost with what you're suggesting. The way I would think of it would be to save the copyright information locally until the tab is refreshed or goes to a new page.

No, I meant fetching track info once per a new video, and store this info in memory. These are mostly technical details.

**Comment by Wolfrum5150** [comment]
Okay, I was able to test out the preview build a little bit and didn't have any issues with it. I would agree that Method B would be much prefered to Method A, but considering that it requires a rework of the connector, I understand that this isn't a high-priority feature. I'll try to test out the preview build more over the next few days and I'll get back to you on any unexpected behaviors.

**Comment by Wolfrum5150** [comment]
After a bit of searching specifically through live and album videos, I was able to find a possible oversight we had when initially discussing the hierarchy of metadata. 

In [this album stream](https://www.youtube.com/watch?v=2aK3hF3bVPs), the preview build works as intended, ignoring the multiple track listings and immediately attempting to read data from the title and description.

There are not any chapters or timestamps for individual songs in the stream, so Web Scrobbler is unable to extract the individual songs of that album and continues as if it were a single song.

**However, since Web Scrobbler assumes that this video is a single song, the album is listed as the artist and the artist as the track.**
![image](https://user-images.githubusercontent.com/79181893/109224944-a2799000-778a-11eb-92a3-093f2e879219.png)

It might be useful to use multiple copyright listings and a lack of chapters as a sign to reverse that order, as most albums are listed as **Band - Album.**

This idea is just in its rough state, though, so it will obviously need to be refined in order to be coded. I would like to hear your thoughts on this feature.

**Comment by alexesprit** [comment]
> However, since Web Scrobbler assumes that this video is a single song, the album is listed as the artist and the artist as the track.
> It might be useful to use multiple copyright listings and a lack of chapters as a sign to reverse that order, as most albums are listed as Band - Album.

That's odd. This is how it was recognized for me (I tested this on the same build):
![image](https://user-images.githubusercontent.com/1119267/109266521-3bef8300-7819-11eb-93e0-19a37e7709a0.png)

**Comment by alexesprit** [comment]
I see you edited the information for this video. Could you revert your edited info and play video again?

**Comment by Wolfrum5150** [comment]
Okay, now I got the same format as you. I don't remember changing the format at all except to see why it was backwards. 
Guess it was just user error.
![image](https://user-images.githubusercontent.com/79181893/109267703-fe1e3a80-77d7-11eb-8ae4-564842b67af4.png)

**Comment by alexesprit** [comment]
@Wolfrum5150 Could you test this build?

[web-scrobbler-chrome.zip](https://github.com/web-scrobbler/web-scrobbler/files/6054129/web-scrobbler-chrome.zip)
[web-scrobbler-firefox.zip](https://github.com/web-scrobbler/web-scrobbler/files/6054130/web-scrobbler-firefox.zip)

**Comment by Wolfrum5150** [comment]
### TL;DR Something weird is going on with ads before videos

I got some very odd behavior for this build. When attempting to use the build on the [Destiny Potato album](https://www.youtube.com/watch?v=2aK3hF3bVPs), I got this result when forcing track recognition: 
![image](https://user-images.githubusercontent.com/79181893/109400031-50fd0c80-7914-11eb-9a79-793cfb1a97db.png)

It looks like it took the channel name for the artist and the ad the played before the video as the song name.

The channel name would've been fine if this group didn't use the same channel for their old and new band.
As for the video ad, this is clearly incorrect. refreshing the page resulted in a different ad and gave me this track name instead. 
![image](https://user-images.githubusercontent.com/79181893/109400118-dd0f3400-7914-11eb-8def-45f47c3cfb53.png)

Using the [RHCP video I sent before](https://www.youtube.com/watch?v=Z0AXjUy1_gY), the extension worked flawlessly. The description was closed as normal and the scrobble showed the copyright information as expected. There was no ad on this video.
![image](https://user-images.githubusercontent.com/79181893/109400217-6aeb1f00-7915-11eb-9b45-5d24930b434d.png)

I figured that it was an issue with the ad loading before a video, so I tried finding a previous video with a single track in the description with an ad before it to test that behavior. [This Haken video I listed before](https://www.youtube.com/watch?v=PMzgh_1j838) had an ad played before the video, but it returned with the scrobble as expected, even noting that I played this song before from my streams of it on Spotify. 
![image](https://user-images.githubusercontent.com/79181893/109400327-22803100-7916-11eb-9f32-d3a36d39878a.png)

All that left was a track with multiple copyright tracks listed, but without chapters, with an ad playing before it. It took my a while to get an ad to play on such an album, but after searching for a while I was able to get this scrobbler from [this video:](https://www.youtube.com/watch?v=WF0k6Z9kftk) 
![image](https://user-images.githubusercontent.com/79181893/109400694-0f6e6080-7918-11eb-8bf6-8850b55f06fd.png)

So it seems to be that something goes wrong with ads playing when using the new method on a song that has multiple track listings. My guess is that chapters eliminate this problem since they work differently, but I did not get around to testing that.

**Comment by alexesprit** [comment]
Interesting. I also changed code related to ads recognition, but it's expected to work in the same way. I've just tested that build, and have no issues you described.

Did the extension recognize ads before?

**Comment by Wolfrum5150** [comment]
No, this is the first time that it has recognized ads.

I was using a fresh install of Firefox Developer Edition with no other extensions installed to simulate normal conditions. Normally, I use uBlock Origin, so I don't get ads on Youtube. But this is the first time I am getting this type of behavior. Here are the Web Scrobbler settings I used when testing:
![image](https://user-images.githubusercontent.com/79181893/109403333-9a585680-792a-11eb-91d0-273da3b0f465.png)
There are no edited scrobbles in my cache so it's not like I'm changing anything either.

Any time I was able to find a track with an ad and multiple listings but no chapters, I was able to recreate the issue.

If it is a Firefox specific issue, this is the build that I used to test the extension (I am unable to use unsigned extensions on the normal version of Firefox):
![image](https://user-images.githubusercontent.com/79181893/109403383-12268100-792b-11eb-9e71-15728d3cbda0.png)

**Comment by alexesprit** [comment]
Not sure if it's Firefox-specific issue. I installed the same version, and I can't reproduce the issue.

Could you also test these ones?

<details>
  <summary><b>Builds</b> (Click to expand)</summary>

[web-scrobbler-firefox+debug.zip](https://github.com/web-scrobbler/web-scrobbler/files/6056419/web-scrobbler-firefox%2Bdebug.zip)
[web-scrobbler-firefox+new-check.zip](https://github.com/web-scrobbler/web-scrobbler/files/6056420/web-scrobbler-firefox%2Bnew-check.zip)
[web-scrobbler-firefox+revert-check.zip](https://github.com/web-scrobbler/web-scrobbler/files/6056421/web-scrobbler-firefox%2Brevert-check.zip)
</details>

Short description of these builds:
1. `debug` is the same as the previous build, but contains some additional debug info.
2. `new-check` contains updated check for ads.
3. `revert-check` contains reverted check for ads.

How to get additional debug log (**for `debug` build only**):
1. Open the developer console (Ctrl+Shift+I), and select the "Console" tab.
2. Keep the console opened, and find any video with ad.
3. While the ad is playing, the extension will write debug log like on the screenshot below.
4. Expand the row that was added when ad was playing and send me (you can copy/paste it or send as screenshot).

**One small note**: this debug info is added after the ad check is executed. If you see no debug info while the ad is playing, but the ad is recognized by the extension, it means the ad check works fine and the problem is something else.

<details>
  <summary><b>Screenshot</b> (Click to expand)</summary>

![image](https://user-images.githubusercontent.com/1119267/109412271-0fbe3880-79b8-11eb-8f68-8a54ef41c5b5.png)
</details>

**Comment by Wolfrum5150** [comment]
Using the debug window, I wasn't able to get the isPlaying state to change to true, but it did change to false in this image:

**Comment by Wolfrum5150** [comment]
I would also like to point out that I do not currently get this error with the current build published on Firefox. Using that video, removing uBlock so that ads and load normally, I still do not get the change to true, but the information is correct:
![image](https://user-images.githubusercontent.com/79181893/110860957-70d4ee80-828b-11eb-9baf-98531c5b0bd1.png)

**Comment by alexesprit** [comment]
Interesting. For some reason you have no "ad-showing" class in the DOMTokenList.

**Comment by Wolfrum5150** [comment]
Well when it scrobbles, it doesn't actually start scrobbling until the ad is completed. For the length of the ad, it shows that Web Scrobbler is waiting for Youtube to start playing a video. It's only after the actual video starts when Web Scrobbler incorrectly scrobbles. The only issues I could think of is that it's a Firefox Developer Edition specific issue, or it has something to do with settings in Firefox, since it has built in privacy settings.

Here's is what my privacy settings are set to:

**Comment by Wolfrum5150** [comment]
Out of curiosity, I tried using the live build on Firefox Developer Edition and I was able to get the correct scrobbling information:
![image](https://user-images.githubusercontent.com/79181893/111042786-106dba80-840d-11eb-8d31-f696720c18ab.png)

It seems to be something specifically with this new build that is causing the issue with the ads since all of the test builds had this issue while the live version (without the new check) did not have this issue.
